pkg/api/server/server.go

```diff
@@ -0,0 +1,127 @@
+package server
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/els0r/goProbe/pkg/api"
+	"github.com/gin-gonic/gin"
+)
+
+type Option func(*DefaultServer)
+
+// DefaultServer is the default API server, allowing middlewares and settings to be
+// re-used across binaries serving an API
+type DefaultServer struct {
+	// api handling
+	// TODO: authorize API access
+	keys []string
+
+	debug bool
+
+	// telemetry
+	profiling bool
+	metrics   bool
+
+	addr string
+
+	srv    *http.Server
+	router *gin.Engine
+
+	unixSocketFile string
+}
+
+// WithDebugMode runs the gin server in debug mode (e.g. not setting the release mode)
+func WithDebugMode(enabled bool) Option {
+	return func(server *DefaultServer) {
+		server.debug = enabled
+	}
+}
+
+// WithProfiling enables runtime profiling endpoints
+func WithProfiling(enabled bool) Option {
+	return func(server *DefaultServer) {
+		server.profiling = enabled
+	}
+}
+
+// WithMetrics enables prometheus metrics endpoints
+func WithMetrics(enabled bool) Option {
+	return func(server *DefaultServer) {
+		server.metrics = enabled
+	}
+}
+
+// NewDefault creates a new API server
+func NewDefault(addr string, opts ...Option) *DefaultServer {
+	s := &DefaultServer{
+		addr: addr,
+	}
+
+	router := gin.New()
+	router.MaxMultipartMemory = 32 << 20 // 32 MiB
+
+	router.Use(gin.Recovery())
+
+	// make sure that unix sockets are handled if they are provided
+	s.unixSocketFile = api.ExtractUnixSocket(addr)
```

Comment:
Does this need sanitization / errror handling maybe? What if BS is provided / specified?

Reply:
Not needed. If BS is provided, `server.unixSocketFile` will be empty, meaning the server will try to mount a tcp listener and then rightfully complain that it's BS.